### PR TITLE
sparse gp regression cookbook

### PR DESCRIPTION
**Pull request by Saurabh7**

### CODE REVIEW

**Comment by Saurabh7** [comment]
@karlnapf  Pls have a look and let me know what else could be added.

**Comment by karlnapf** [comment]
@yorkerlin can you comment here as well, see the second check for a html rendering of the page

File: doc/cookbook/source/examples/gaussian_processes/sparse_gaussian_process_regression.rst [hunk 1]
```diff
@@ -0,0 +1,49 @@
+==================================
+Sparse Gaussian Process Regression
```
Comment (karlnapf) [inline doc/cookbook/source/examples/gaussian_processes/sparse_gaussian_process_regression.rst]: Sparse Gaussian Process Approximations

File: doc/cookbook/source/examples/gaussian_processes/sparse_gaussian_process_regression.rst [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+==================================
+
+
+A Gaussian Process is a method for bayesian non-linear non-parametric regression and classification. Sparse approximations for full GP are done to reduce computational scaling. This requires selecting an additional :math:`m` latent variables which could be a  subset of training points. This subset form a pseudo data set with inputs :math:`\mathbf{X}'` and targets :math:`\mathbf{f}'` .
```
Comment (karlnapf) [inline doc/cookbook/source/examples/gaussian_processes/sparse_gaussian_process_regression.rst]: Just refer to the GP cookbook for saying what a GP is

Comment (karlnapf) [inline doc/cookbook/source/examples/gaussian_processes/sparse_gaussian_process_regression.rst]: The inducing points can either be a subset of the training data or being optimized free form

File: doc/cookbook/source/examples/gaussian_processes/sparse_gaussian_process_regression.rst [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+Given the training data, the predictive distribution :math:`y^*` for a new input point :math:`\mathbf{x}^*` will then be:
+
+.. math::
+	p(y^*|\mathbf{x}^*, \mathbf{y}, \mathbf{X}, \mathbf{X}')=\int p(\mathbf{y}^*|\mathbf{x}^*, \mathbf{X}',\mathbf{f}')p(\mathbf{f}'| \mathbf{y}, \mathbf{X}, \mathbf{X}')df'
+
+See :cite:`Rasmussen2005GPM` for a comprehensive treatment of Gaussian Processes (see Chapter 2 for regression).
```
Comment (karlnapf) [inline doc/cookbook/source/examples/gaussian_processes/sparse_gaussian_process_regression.rst]: Need a better reference, i.e. http://www.jmlr.org/papers/volume6/quinonero-candela05a/quinonero-candela05a.pdf

File: doc/cookbook/source/examples/gaussian_processes/sparse_gaussian_process_regression.rst [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+Imagine we have files with training and test data. We create `CDenseFeatures` (here 64 bit floats aka RealFeatures) and :sgclass:`CRegressionLabels` as:
+
+.. sgexample:: sparse_gaussian_process_regression.sg:create_features
+
+To fit the input (training) data :math:`\mathbf{X}`, we have to choose an appropriate :sgclass:`CMeanFunction` and  :sgclass:`CKernel` and instantiate them. Here we use a basic :sgclass:`CZeroMean` and a :sgclass:`CGaussianKernel` with chosen width parameter.
+
+.. sgexample::  sparse_gaussian_process_regression.sg:create_appropriate_kernel_and_mean_function
+
+We need to specify the inference method to find the posterior distribution of the function values :math:`\mathbf{f}`. Here we choose to perform the Fully independent Training Conditional (FITC) inference with an instance of :sgclass:`CFITCInferenceMethod`. We use another feature instance for inducing points and add a simple subset for demonstration. The inference method is then created and we pass it the chosen kernel, the training features, the mean function, the labels, an instance of :sgclass:`CGaussianLikelihood` and the inducing features.
```
Comment (karlnapf) [inline doc/cookbook/source/examples/gaussian_processes/sparse_gaussian_process_regression.rst]: FITC is a crappy approximation, I suggest we use the Titias here?
@yorkerlin ?

File: examples/meta/src/gaussian_processes/sparse_gaussian_process_regression.sg [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+#![create_appropriate_kernel_and_mean_function]
+real width = 1
+GaussianKernel kernel(features_train, features_train, width)
+ZeroMean mean_function()
+#![create_appropriate_kernel_and_mean_function]
+
+#![create_inference]
+IntVector inducing_points(3)
+inducing_points[0]=0
```
Comment (karlnapf) [inline examples/meta/src/gaussian_processes/sparse_gaussian_process_regression.sg]: can't we learn the inducing points and only specify their number here?
@yorkerlin whats the cleanest interface way to do this?

File: doc/cookbook/source/examples/gaussian_processes/sparse_gaussian_process_regression.rst (hunk 4 above)
Comment (yorkerlin) [inline doc/cookbook/source/examples/gaussian_processes/sparse_gaussian_process_regression.rst]: Yes. We can use `VarDTCInferenceMethod`

File: examples/meta/src/gaussian_processes/sparse_gaussian_process_regression.sg (hunk 5 above)
Comment (yorkerlin) [inline examples/meta/src/gaussian_processes/sparse_gaussian_process_regression.sg]: In practice, we can first group training points into K clusters using unsupervised algorithms (eg, PCA). Then centers of K clusters are used as K inducing points.

Comment (yorkerlin) [inline examples/meta/src/gaussian_processes/sparse_gaussian_process_regression.sg]: We could also randomly select K training points as initial inducing points. Then we optimize the marginal likelihood to good inducing points.

File: doc/cookbook/source/examples/gaussian_processes/sparse_gaussian_process_regression.rst (hunk 3 above)
Comment (yorkerlin) [inline doc/cookbook/source/examples/gaussian_processes/sparse_gaussian_process_regression.rst]: I second that.

Comment (yorkerlin) [inline doc/cookbook/source/examples/gaussian_processes/sparse_gaussian_process_regression.rst]: I second that.

File: doc/cookbook/source/examples/gaussian_processes/sparse_gaussian_process_regression.rst (hunk 1 above)
Comment (yorkerlin) [inline doc/cookbook/source/examples/gaussian_processes/sparse_gaussian_process_regression.rst]: I think this notebook is only for Gaussian likelihood.
 
For other likelihoods (eg, t distribution and Logistic distribution), we should use Laplace method.

Comment (yorkerlin) [inline doc/cookbook/source/examples/gaussian_processes/sparse_gaussian_process_regression.rst]: @Saurabh7 
Please check out the table at http://www.gaussianprocess.org/gpml/code/matlab/doc/ for choosing right inference method for one  likelihood.

**Comment by vigsterkr** [comment]
@Saurabh7 what's the update on this one?

File: examples/meta/src/gaussian_processes/sparse_gaussian_process_regression.sg (hunk 5 above)
Comment (Saurabh7) [inline examples/meta/src/gaussian_processes/sparse_gaussian_process_regression.sg]: I can select the points randomly. Although I cant see a clean way other than adding subset in meta examples.

File: examples/meta/src/gaussian_processes/sparse_gaussian_process_regression.sg [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+#![create_features]
+RealFeatures features_train(f_feats_train)
+RealFeatures features_test(f_feats_test)
+RegressionLabels labels_train(f_labels_train)
+RegressionLabels labels_test(f_labels_test)
+#![create_features]
+
+#![create_appropriate_kernel_and_mean_function]
+real width = 1
```
Comment (sorig) [inline examples/meta/src/gaussian_processes/sparse_gaussian_process_regression.sg]: Please use `1.0`. We currently use the `auto` type for C++ translations, so we need to make sure the literals are correctly typed to get the correct serialization for integration testing.

**Comment by karlnapf** [comment]
What about getting this guy merged soon?

**Comment by karlnapf** [comment]
I'll  merge this and clean it up then

File: examples/meta/src/gaussian_processes/sparse_gaussian_process_regression.sg (hunk 5 above)
Comment (karlnapf) [inline examples/meta/src/gaussian_processes/sparse_gaussian_process_regression.sg]: @yorkerlin would be cool to get this stuff in in a nicer way.... I merged and updated the example, see examples on our website

**Pull request merged**